core/src/main/java/io/grpc/internal/InternalSubchannel.java

```diff
@@ -390,6 +391,18 @@ public void shutdown(Status reason) {
     }
   }
 
+  @Override
+  public String toString() {
+    Object addressGroupCopy;
+    synchronized (lock) {
+      addressGroupCopy = addressGroup.toString();
```

Comment:
I think you don't need call `toString()` here, EAG is immutable and pretty thread safe to call toString out of lock.

Reply:
The field may still be set in one thread but read in another thread, such as the server thread for Channelz.

Reply:
>The field may still be set in one thread but read in another thread, such as the server thread for Channelz.

I don't see any problem with that. When you return the string, the returned value soon becomes stale even if you enclose everything of the toString method with the lock. toString() needn't be in the lock as long as the copying is in the lock
```java
synchronized (lock) {
   addressGroupCopy = addressGroup; 
   // no need to call toString() for a long list of addresses
}
```

Reply:
I see what you mean. Done.